<a href="https://colab.research.google.com/github/fluffymouses/fluffymouses/blob/main/lab1/%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [1]:
import pandas as pd

In [3]:
df = pd.read_csv("/content/sq/crimes_against_women_2001-2014.csv")
df

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
0               0  ANDHRA PRADESH           ADILABAD  2001    50   
1               1  ANDHRA PRADESH          ANANTAPUR  2001    23   
2               2  ANDHRA PRADESH           CHITTOOR  2001    27   
3               3  ANDHRA PRADESH           CUDDAPAH  2001    20   
4               4  ANDHRA PRADESH      EAST GODAVARI  2001    23   
...           ...             ...                ...   ...   ...   
10672         832     Lakshadweep        Lakshadweep  2014     1   
10673         833     Lakshadweep  Total District(s)  2014     1   
10674         834      Puducherry           Karaikal  2014     3   
10675         835      Puducherry         Puducherry  2014     7   
10676         836      Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
0                            30            16   
1                            30             7   
2                            34            14   
3                            20            17   
4                            26            12   
...                         ...           ...   
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
0                                                    149     
1                                                    118     
2                                                    112     
3                                                    126     
4                                                    109     
...                                                  ...     
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                              34                                  175   
1                              24                                  154   
2                              83                                  186   
3                              38                                   57   
4                              58                                  247   
...                           ...                                  ...   
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  
...                     ...  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0  

[10677 rows x 11 columns]

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [5]:
df.columns

Index(['Unnamed: 0', 'STATE/UT', 'DISTRICT', 'Year', 'Rape',
       'Kidnapping and Abduction', 'Dowry Deaths',
       'Assault on women with intent to outrage her modesty',
       'Insult to modesty of Women', 'Cruelty by Husband or his Relatives',
       'Importation of Girls'],
      dtype='object')

In [6]:
df.head()

Unnamed: 0        STATE/UT       DISTRICT  Year  Rape  \
0           0  ANDHRA PRADESH       ADILABAD  2001    50   
1           1  ANDHRA PRADESH      ANANTAPUR  2001    23   
2           2  ANDHRA PRADESH       CHITTOOR  2001    27   
3           3  ANDHRA PRADESH       CUDDAPAH  2001    20   
4           4  ANDHRA PRADESH  EAST GODAVARI  2001    23   

   Kidnapping and Abduction  Dowry Deaths  \
0                        30            16   
1                        30             7   
2                        34            14   
3                        20            17   
4                        26            12   

   Assault on women with intent to outrage her modesty  \
0                                                149     
1                                                118     
2                                                112     
3                                                126     
4                                                109     

   Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                          34                                  175   
1                          24                                  154   
2                          83                                  186   
3                          38                                   57   
4                          58                                  247   

   Importation of Girls  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0

In [7]:
df.tail()

Unnamed: 0     STATE/UT           DISTRICT  Year  Rape  \
10672         832  Lakshadweep        Lakshadweep  2014     1   
10673         833  Lakshadweep  Total District(s)  2014     1   
10674         834   Puducherry           Karaikal  2014     3   
10675         835   Puducherry         Puducherry  2014     7   
10676         836   Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0

In [8]:
df.describe()

Unnamed: 0          Year          Rape  Kidnapping and Abduction  \
count  10677.000000  10677.000000  10677.000000              10677.000000   
mean    3871.570198   2007.697949     57.989885                 69.888358   
std     2816.369918      4.046874    214.230398                311.623450   
min        0.000000   2001.000000      0.000000                  0.000000   
25%     1009.000000   2004.000000      8.000000                  6.000000   
50%     3678.000000   2008.000000     22.000000                 20.000000   
75%     6347.000000   2011.000000     44.000000                 49.000000   
max     9016.000000   2014.000000   5076.000000              10626.000000   

       Dowry Deaths  Assault on women with intent to outrage her modesty  \
count  10677.000000                                       10677.000000     
mean      20.181699                                         113.539196     
std       98.276531                                         458.903951     
min        0.000000                                           0.000000     
25%        1.000000                                          10.000000     
50%        5.000000                                          34.000000     
75%       16.000000                                          85.000000     
max     2469.000000                                       10001.000000     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
count                10677.000000                         10677.000000   
mean                    27.419313                           209.224314   
std                    167.806797                           905.664362   
min                      0.000000                             0.000000   
25%                      0.000000                            11.000000   
50%                      2.000000                            50.000000   
75%                     12.000000                           144.000000   
max                   4970.000000                         23278.000000   

       Importation of Girls  
count          10677.000000  
mean               0.175330  
std                2.228637  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max               83.000000

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета





1.   Unnamed: 0                                             3871.570198
Year                                                   2007.697949
Rape                                                     57.989885
Kidnapping and Abduction                                 69.888358
Dowry Deaths                                             20.181699
Assault on women with intent to outrage her modesty     113.539196
Insult to modesty of Women                               27.419313
Cruelty by Husband or his Relatives                     209.224314
Importation of Girls                                      0.175330
2.   Min:
STATE/UT                                                   A & N ISLANDS
DISTRICT                                               24 PARGANAS NORTH
Year                                                              2001.0
Rape                                                                   0
Kidnapping and Abduction                                               0
Dowry Deaths                                                           0
Assault on women with intent to outrage her modesty                    0
Insult to modesty of Women                                             0
Cruelty by Husband or his Relatives                                    0
Importation of Girls                                                   0
  Max:
Unnamed: 0                                                    9016
STATE/UT                                               West Bengal
DISTRICT                                                 Zunheboto
Year                                                        2014.0
Rape                                                          5076
Kidnapping and Abduction                                     10626
Dowry Deaths                                                  2469
Assault on women with intent to outrage her modesty          10001
Insult to modesty of Women                                    4970
Cruelty by Husband or his Relatives                          23278
Importation of Girls                                            83

##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [9]:
df.Year = df.Year.astype("Float64")
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10677 entries, 0 to 10676
Data columns (total 11 columns):
 #   Column                                               Non-Null Count  Dtype  
---  ------                                               --------------  -----  
 0   Unnamed: 0                                           10677 non-null  int64  
 1   STATE/UT                                             10677 non-null  object 
 2   DISTRICT                                             10677 non-null  object 
 3   Year                                                 10677 non-null  Float64
 4   Rape                                                 10677 non-null  int64  
 5   Kidnapping and Abduction                             10677 non-null  int64  
 6   Dowry Deaths                                         10677 non-null  int64  
 7   Assault on women with intent to outrage her modesty  10677 non-null  int64  
 8   Insult to modesty of Women                           10677 non-nul

Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [10]:
df[df["Year"] > 2013]

Unnamed: 0        STATE/UT           DISTRICT    Year  Rape  \
9840            0  Andhra Pradesh          Anantapur  2014.0    35   
9841            1  Andhra Pradesh           Chittoor  2014.0    32   
9842            2  Andhra Pradesh           Cuddapah  2014.0    28   
9843            3  Andhra Pradesh      East Godavari  2014.0    85   
9844            4  Andhra Pradesh   Guntakal Railway  2014.0     0   
...           ...             ...                ...     ...   ...   
10672         832     Lakshadweep        Lakshadweep  2014.0     1   
10673         833     Lakshadweep  Total District(s)  2014.0     1   
10674         834      Puducherry           Karaikal  2014.0     3   
10675         835      Puducherry         Puducherry  2014.0     7   
10676         836      Puducherry  Total District(s)  2014.0    10   

       Kidnapping and Abduction  Dowry Deaths  \
9840                        106            25   
9841                         34            17   
9842                         16            16   
9843                         25             7   
9844                          0             0   
...                         ...           ...   
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
9840                                                 436     
9841                                                 135     
9842                                                 215     
9843                                                 519     
9844                                                   0     
...                                                  ...     
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
9840                           26                                  165   
9841                           94                                  278   
9842                           12                                   91   
9843                           62                                  464   
9844                            0                                    0   
...                           ...                                  ...   
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
9840                      0  
9841                      0  
9842                      0  
9843                      0  
9844                      0  
...                     ...  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0  

[837 rows x 11 columns]

In [13]:
df[df["Rape"] > 23]

Unnamed: 0        STATE/UT           DISTRICT    Year  Rape  \
0               0  ANDHRA PRADESH           ADILABAD  2001.0    50   
2               2  ANDHRA PRADESH           CHITTOOR  2001.0    27   
6               6  ANDHRA PRADESH             GUNTUR  2001.0    54   
7               7  ANDHRA PRADESH     HYDERABAD CITY  2001.0    37   
8               8  ANDHRA PRADESH         KARIMNAGAR  2001.0    56   
...           ...             ...                ...     ...   ...   
10664         824        Delhi UT              South  2014.0   280   
10665         825        Delhi UT         South-East  2014.0   266   
10666         826        Delhi UT         South-West  2014.0   234   
10670         830        Delhi UT               West  2014.0   251   
10671         831        Delhi UT  Total District(s)  2014.0  2096   

       Kidnapping and Abduction  Dowry Deaths  \
0                            30            16   
2                            34            14   
6                            51             7   
7                            39            24   
8                            49            62   
...                         ...           ...   
10664                       369             6   
10665                       402            24   
10666                       397            17   
10670                       546            13   
10671                      4034           153   

       Assault on women with intent to outrage her modesty  \
0                                                    149     
2                                                    112     
6                                                    139     
7                                                    118     
8                                                    414     
...                                                  ...     
10664                                                595     
10665                                                469     
10666                                                494     
10670                                                484     
10671                                               4322     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                              34                                  175   
2                              83                                  186   
6                             129                                  378   
7                              27                                  746   
8                              81                                  224   
...                           ...                                  ...   
10664                         665                                  143   
10665                         123                                  312   
10666                          68                                  318   
10670                          63                                  416   
10671                        1361                                 3173   

       Importation of Girls  
0                         0  
2                         0  
6                         0  
7                         0  
8                         0  
...                     ...  
10664                     0  
10665                     0  
10666                     0  
10670                     0  
10671                     0  

[5036 rows x 11 columns]

In [16]:
df[df["DISTRICT"] == "CHITTOOR"]

Unnamed: 0        STATE/UT  DISTRICT    Year  Rape  \
2              2  ANDHRA PRADESH  CHITTOOR  2001.0    27   
718          718  ANDHRA PRADESH  CHITTOOR  2002.0    28   
1437        1437  ANDHRA PRADESH  CHITTOOR  2003.0    36   
2165        2165  ANDHRA PRADESH  CHITTOOR  2004.0    28   
2894        2894  ANDHRA PRADESH  CHITTOOR  2005.0    35   
3627        3627  ANDHRA PRADESH  CHITTOOR  2006.0    31   
4367        4367  ANDHRA PRADESH  CHITTOOR  2007.0    31   
5110        5110  ANDHRA PRADESH  CHITTOOR  2008.0    30   
5871        5871  ANDHRA PRADESH  CHITTOOR  2009.0    33   
6638        6638  ANDHRA PRADESH  CHITTOOR  2010.0    44   
7417        7417  ANDHRA PRADESH  CHITTOOR  2011.0    17   
8208        8208  ANDHRA PRADESH  CHITTOOR  2012.0    28   
9019           2  Andhra Pradesh  CHITTOOR  2013.0    31   

      Kidnapping and Abduction  Dowry Deaths  \
2                           34            14   
718                         39            16   
1437                        34            23   
2165                        27            22   
2894                        58             9   
3627                        37            18   
4367                        47            25   
5110                        57            26   
5871                        46            19   
6638                        56            32   
7417                        47            25   
8208                        49            33   
9019                        27            13   

      Assault on women with intent to outrage her modesty  \
2                                                   112     
718                                                 126     
1437                                                121     
2165                                                116     
2894                                                 91     
3627                                                123     
4367                                                118     
5110                                                104     
5871                                                120     
6638                                                136     
7417                                                124     
8208                                                 90     
9019                                                119     

      Insult to modesty of Women  Cruelty by Husband or his Relatives  \
2                             83                                  186   
718                           40                                  188   
1437                         173                                   60   
2165                         197                                  234   
2894                         223                                  202   
3627                         190                                  249   
4367                         197                                  395   
5110                         250                                  288   
5871                         268                                  333   
6638                         121                                  388   
7417                         168                                  421   
8208                         269                                  339   
9019                          84                                  435   

      Importation of Girls  
2                        0  
718                      0  
1437                     0  
2165                     0  
2894                     0  
3627                     0  
4367                     0  
5110                     0  
5871                     0  
6638                     0  
7417                     0  
8208                     0  
9019                     0

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [18]:
df.groupby(["Importation of Girls"]).mean()

<ipython-input-18-68afc2ff3e6a>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby(["Importation of Girls"]).mean()


Unnamed: 0         Year         Rape  \
Importation of Girls                                          
0                     3867.384172  2007.703722    48.174440   
1                     3737.386792  2007.037736   300.113208   
2                     4003.617021  2008.978723   319.042553   
3                     4681.708333     2007.375   417.333333   
4                     4798.062500    2007.9375   241.250000   
5                     4380.000000  2008.611111   787.277778   
6                     4633.666667       2007.0   845.166667   
7                      713.000000       2006.6  1282.400000   
8                     4679.285714       2007.0   576.571429   
9                     5014.428571  2010.428571   271.857143   
10                    7711.666667       2011.0   353.000000   
11                    2857.000000       2004.0   108.000000   
12                    5954.142857  2008.428571   744.142857   
14                     573.000000       2001.0   423.000000   
15                    5221.000000       2008.0    32.000000   
21                    1548.000000       2003.0    25.000000   
22                    4855.500000       2007.5   657.000000   
23                    5981.000000       2009.0    18.000000   
25                    3011.500000       2005.0    37.000000   
27                    4556.000000       2007.0    57.500000   
29                     107.000000       2001.0    19.000000   
31                    5991.000000       2009.0   929.000000   
32                    8536.000000       2012.0   621.000000   
34                    1690.500000       2003.0    14.000000   
35                    2281.000000       2004.0  1390.000000   
36                    1695.500000       2003.0   797.000000   
37                    1553.000000       2003.0   985.000000   
38                     831.000000       2002.0  1040.000000   
39                    2740.000000       2004.5   409.000000   
42                    3746.000000       2006.0  1232.000000   
45                    7810.000000       2011.0  3406.000000   
48                    3005.000000       2005.0    28.000000   
56                    4486.000000       2007.0  1555.000000   
60                    3590.000000       2005.0   148.000000   
61                    3597.000000       2005.0  1686.000000   
74                    3013.000000       2005.0  1147.000000   
83                     115.000000       2001.0   888.000000   

                      Kidnapping and Abduction  Dowry Deaths  \
Importation of Girls                                           
0                                    59.126166     16.369843   
1                                   305.216981     69.867925   
2                                   416.829787     71.638298   
3                                   401.750000    183.416667   
4                                   593.437500    141.812500   
5                                   815.222222    300.833333   
6                                   391.000000    237.833333   
7                                  1630.800000    489.000000   
8                                   873.571429    306.714286   
9                                   606.285714     76.285714   
10                                 1030.333333    478.333333   
11                                   80.000000     30.000000   
12                                  928.571429    247.428571   
14                                  607.000000    191.000000   
15                                   62.000000     60.000000   
21                                   18.500000     28.000000   
22                                  916.000000    620.000000   
23                                   11.000000     41.000000   
25                                   18.500000     36.500000   
27                                   36.500000     29.500000   
29                                   29.000000     11.000000   
31                                 1986.000000   1295.000000   
32                           

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [ ]:
#Агрегация